src/test/timedata_tests.cpp

```diff
@@ -34,4 +38,58 @@ BOOST_AUTO_TEST_CASE(util_MedianFilter)
     BOOST_CHECK_EQUAL(filter.median(), 7);
 }
 
+void MultiAddTimeData(int n, int64_t offset)
+{
+    static int cnt = 0;
+    for (int i = 0; i < n; ++i) {
+        CNetAddr addr;
+        addr.SetInternal(std::to_string(++cnt));
+        AddTimeData(addr, offset);
+    }
+}
+
+
+BOOST_AUTO_TEST_CASE(addtimedata)
+{
+    BOOST_CHECK_EQUAL(GetTimeOffset(), 0);
+
+    //Part 1: Add large offsets to test a warning message that our clock may be wrong.
+    MultiAddTimeData(3, DEFAULT_MAX_TIME_ADJUSTMENT + 1); //filter size 4, initialized with 0
+
+    noui_suppress();
+    MultiAddTimeData(1, DEFAULT_MAX_TIME_ADJUSTMENT + 1); //filter size 5
+    noui_reconnect();
+
+    std::string warning = GetWarnings("gui");
+    BOOST_CHECK(warning.find("clock is wrong") != std::string::npos);
+
+    // nTimeOffset is not changed if the median of offsets exceeds DEFAULT_MAX_TIME_ADJUSTMENT
+    BOOST_CHECK_EQUAL(GetTimeOffset(), 0);
+
+    // Part 2: Test positive and negative medians by adding more offsets
+    MultiAddTimeData(4, 100); // filter size 9
+    BOOST_CHECK_EQUAL(GetTimeOffset(), 100);
+    MultiAddTimeData(10, -100); //filter size 19
+    BOOST_CHECK_EQUAL(GetTimeOffset(), -100);
+
+    // Part 3: Test behaviour when filter has reached maximum number of offsets
+    const int MAX_SAMPLES = 200;
+    int nfill = (MAX_SAMPLES - 3 - 19) / 2; //89
+    MultiAddTimeData(nfill, 100);
+    MultiAddTimeData(nfill, -100); //filter size MAX_SAMPLES - 3
+    BOOST_CHECK_EQUAL(GetTimeOffset(), -100);
+
+    MultiAddTimeData(2, 100); //filter size MAX_SAMPLES -1
+    BOOST_CHECK_EQUAL(GetTimeOffset(), 0);
+
+    // After the number of offsets has reached MAX_SAMPLES -1 (=199), nTimeOffset will never change
+    // because it is only updated when the number of elements in the filter becomes odd. It was decided
+    // not to fix this because it prevents possible attacks. See the comment in AddTimeData() or issue #4521
+    // for a more detailed explanation.
+    MultiAddTimeData(MAX_SAMPLES, 200); // filter median is 200, but nTimeOffset will not change
```

Comment:
Shouldn't be better to try to increment by one instead of MAX_SAMPLES to see if threshold effect happens as expected?

Reply:
I wanted to point out that you can replace as much elements in `vTimeOffsets`  as you want without `nTimeOffset` ever changing, but yeah, MAX_SAMPLES is probably a bit overkill, so will change this to a lower value.

Reply:
Yes, median doesn't change anymore at MAX_SAMPLES so that's this threshold effect you're interested to see I think (in this case nothing should change)